src/main/java/fopbot/Direction.java

```
@@ -1,23 +1,155 @@
+     */
+    final int dx;
+    /**
+     * The y offset of the {@link Direction}.
+     */
+    final int dy;
+
+    /**
+     * Creates a new direction.
+     *
+     * @param dx The x-offset of the {@link Direction}.
+     * @param dy The y-offset of the {@link Direction}.
+     */
+    Direction(final int dx, final int dy) {
+        this.dx = dx;
+        this.dy = dy;
+    }
+
+    /**
+     * Gets the x offset of the {@link Direction}.
+     *
+     * @return The x offset of the {@link Direction}.
+     */
+    public int getDx() {
+        return dx;
+    }
+
+    /**
+     * Gets the y offset of the {@link Direction}.
+     *
+     * @return The y offset of the {@link Direction}.
+     */
+    public int getDy() {
+        return dy;
+    }
+
+    /**
+     * Checks whether this direction is a horizontal direction.
+     *
+     * @return {@code true} if this direction is a horizontal direction.
+     */
+    public boolean isHorizontal() {
+        return dy == 0;
+    }
+
+    /**
+     * Checks whether this direction is a vertical direction.
+     *
+     * @return {@code true} if this direction is a vertical direction.
+     */
+    public boolean isVertical() {
+        return dx == 0;
+    }
+
+    /**
+     * Calculates the length of a given vector of an arbitrary dimension.
+     *
+     * @param components the components of the vector.
+     * @return the length of the vector.
+     */
+    private double vectorLength(final double... components) {
+        return Math.sqrt(Arrays.stream(components).map(c -> Math.pow(c, 2)).sum());
+    }
+
+    /**
+     * Returns the direction who best matches the given vector.
+     *
+     * @param x The x-component of the vector.
+     * @param y The y-component of the vector.
+     * @return Returns the direction who best matches the given vector.
+     */
+    public Direction fromVector(final double x, final double y) {
+        if (x == 0 && y == 0) {
+            return NONE;
```

This should throw an error instead of returning an invalid direction

---

will be removed